src/streams.c

```diff
@@ -185,53 +180,66 @@ Obj READ_ALL_COMMANDS(Obj instream, Obj echo, Obj capture, Obj resultCallback)
         return Fail;
     }
 
-    resultList = NEW_PLIST(T_PLIST, 16);
+    volatile Obj resultList = NEW_PLIST(T_PLIST, 16);
+    BOOL rethrow = FALSE;
 
-    do {
-        if (outstream) {
-            // Clean in case there has been any output
-            SET_LEN_STRING(outstreamString, 0);
-        }
+    GAP_TRY
+    {
+        ExecStatus status;
+        do {
+            if (outstream) {
+                // Clean in case there has been any output
+                SET_LEN_STRING(outstreamString, 0);
+            }
 
-        status = ReadEvalCommand(0, &input, &evalResult, &dualSemicolon);
+            UInt dualSemicolon;
+            Obj  evalResult;
 
-        if (!(status & (STATUS_EOF | STATUS_QUIT | STATUS_QQUIT))) {
-            result = NEW_PLIST(T_PLIST, 5);
-            AssPlist(result, 1, False);
-            PushPlist(resultList, result);
+            status = ReadEvalCommand(0, &input, &evalResult, &dualSemicolon);
```

Comment:
After thinking about this PR again, I feel I need to clarify something: `ReadEvalCommand` should never let an exception through, it "swallows" them all and turns any error into a flag in its `status` return value. This is why I did not add `GAP_TRY` to other places in `streams.c` which use `OpenInput`

However, the call to `resultCallback` may throw an exception, and we ought to deal with that (which we now do).